File: 2 JSON Parser/Original Parser Android/src/com/example/javajsonparser/RefType.java
//RefType.java
package com.example.javajsonparser;

import java.io.IOException;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonToken;

public class RefType {
    private static final String NUMBER = "No";
    private static final String NAME = "Name";
    private static final String URL = "Url";
    private static final String LOCATION = "Location";

    private String mNo;
    private String mName;
    private String mUrl;
    private LocationType mLocation;

    public RefType() {
    }

    public RefType(final JsonReader reader) throws IOException {
        deserialize(reader);
    }

    public String getNo() {
        return mNo;
    }

    public void setNo(final String No) {
        this.mNo = No;
    }

    public String getName() {
        return mName;
    }

    public void setName(final String name) {
        this.mName = name;
    }

    public String getUrl() {
        return mUrl;
    }

    public void setUrl(final String url) {
        this.mUrl = url;
    }

    public LocationType getLocation() {
        return mLocation;
    }

    public void setLocation(final LocationType location) {
        this.mLocation = location;
    }

    public void deserialize(final JsonReader reader) throws IOException {
        reader.beginObject();

        while (reader.hasNext()) {
            if (reader.peek() == JsonToken.NAME) {
                String name = reader.nextName();

                if (NAME.equalsIgnoreCase(name)) {
                    reader.beginObject();
                    reader.nextName();
                    this.mName = reader.nextString();
                    reader.endObject();
                }

                if (NUMBER.equalsIgnoreCase(name)) {
                    reader.beginObject();
                    reader.nextName();
                    this.mNo = reader.nextString();
                    reader.endObject();
                }

                if (URL.equalsIgnoreCase(name)) {
                    reader.beginObject();
                    reader.nextName();
                    this.mUrl = reader.nextString();
                    reader.endObject();
                }

                if (LOCATION.equalsIgnoreCase(name)) {
                    this.mLocation = new LocationType(reader);
                }
            } else {
                reader.skipValue();
            }
        }
        reader.endObject();
    }
    
    public String toString(){
    	return this.mName;
    }
}

File: 4 JSON Parser HTTP Cache/C++ Parser+HTTP+Cache iOS w calabash/PocoTest2/RefType.cpp
//RefType.cpp
#include "RefType.h"

RefType::RefType(){
    this->mLocation = 0;
}

RefType::RefType(Object::Ptr object)
{
    this->mLocation = 0;
    deserialize(object);
}

RefType::~RefType(){
    //delete this->mLocation; //Using smart pointer instead is an alternative..
    //this->mLocation = 0;
}


string RefType::getNo(){
    return this->mNo;
}

void RefType::setNo(string no){
    this->mNo = no;
}

string RefType::getName(){
    return this->mName;
}

void RefType::setName(string name){
    this->mName = name;
}

string RefType::getUrl(){
    return this->mUrl;
}

void RefType::setUrl(string url){
    this->mUrl = url;
}

LocationType* RefType::getLocation(){
    return this->mLocation;
}

void RefType::setLocation(LocationType *location){
    this->mLocation = location;
}

void RefType::deserialize(Object::Ptr object){
    if(object->has(NUMBER) && !object->isNull(NUMBER)){
        string str;
        Var no = object->get(NUMBER);
        Object::Ptr subObject = no.extract<Object::Ptr>();
        if (subObject->has("$")&& !subObject->isNull("$")) {
            str = subObject->get("$").convert<string>();
        }else{
            str = no.convert<string>();
        }
        this->mNo = str;
    }
    if(object->has(NAME) && !object->isNull(NAME)){
        string str;
        Var name = object->get(NAME);
        Object::Ptr subObject = name.extract<Object::Ptr>();
        if (subObject->has("$")&& !subObject->isNull("$")) {
            str = subObject->get("$").convert<string>();
        }else{
            str = name.convert<string>();
        }
        this->mName = str;
    }
    if(object->has(URL) && !object->isNull(URL)){
        string str;
        Var url = object->get(URL);
        Object::Ptr subObject = url.extract<Object::Ptr>();
        if (subObject->has("$")&& !object->isNull("$")) {
            str = subObject->get("$").convert<string>();
        }else{
            str = url.convert<string>();
        }
        this->mUrl = str;
    }
    if(object->has(LOCATION) && !object->isNull(LOCATION)){
        Var locationRaw = object->get(LOCATION);
        Object::Ptr locationObject = locationRaw.extract<Object::Ptr>();
        this->mLocation = new LocationType(locationObject);
    }
    else {
        printf("NO! Location");

    }
    
}


